### PR TITLE
nat: fix usage in nat.h of csum.h module

### PR DESCRIPTION
**Pull request by sahid**
csum.h has been initiated to be used for csum update based on protocol that has pseudo header.  In some places in the code of nat.h it has been "hacked" to be used even for other protocol like ICMP that does not have pseudo headers.

### CODE REVIEW

File: bpf/lib/nat.h [hunk 1]
```diff
@@ -602,18 +602,15 @@ static __always_inline int snat_v4_rewrite_egress(struct __ctx_buff *ctx,
...
 						    sizeof(state->to_sport), 0) < 0)
 					return DROP_WRITE_ERROR;
-				from = tuple->sport;
-				to = state->to_sport;
-				flags = 0; /* ICMPv4 has no pseudo-header */
-				sum_l4 = csum_diff(&from, 4, &to, 4, 0);
-				csum.offset = offsetof(struct icmphdr, checksum);
+				if (l4_csum_replace(ctx, off + offsetof(struct icmphdr, checksum),
+						    tuple->sport,
+						    state->to_sport, 0) < 0)
```
Comment (jschwinger233) [inline bpf/lib/nat.h]: Do we need to specify the number of bytes (2) on the lowest four bits of _flags_(the last parameter in bpf_l4_csum_replace)?

Comment (sahid) [reply bpf/lib/nat.h]: Yes Interesting point I'm wondering whether we do right in other places as-well. Let me see...

Comment (sahid) [reply bpf/lib/nat.h]: Thank you for this review I fixed it.

**Comment by sahid** [comment]
Once this is accepted I will update #25051 to make everything coherent.

File: bpf/lib/nat.h [hunk 2]
```diff
@@ -1642,17 +1639,17 @@ static __always_inline int snat_v6_rewrite_egress(struct __ctx_buff *ctx,
...
 			if (ctx_store_bytes(ctx, off +
 					    offsetof(struct icmp6hdr,
 						     icmp6_dataun.u_echo.identifier),
 					    &state->to_sport,
 					    sizeof(state->to_sport), 0) < 0)
 				return DROP_WRITE_ERROR;
-			from = tuple->sport;
-			to = state->to_sport;
-			sum = csum_diff(&from, 4, &to, 4, sum);
+			if (l4_csum_replace(ctx, off + offsetof(struct icmphdr, checksum),
```
Comment (jschwinger233) [inline bpf/lib/nat.h]: `ctx` here holds an ICMPv6 packet, maybe we should use `offsetof(struct icmp6hdr, icmp6_cksum)`?

File: bpf/lib/nat.h [hunk 3]
```diff
@@ -1707,9 +1703,11 @@ static __always_inline int snat_v6_rewrite_ingress(struct __ctx_buff *ctx,
 						    &state->to_dport,
 						    sizeof(state->to_dport), 0) < 0)
 					return DROP_WRITE_ERROR;
-				from = tuple->dport;
-				to = state->to_dport;
-				sum = csum_diff(&from, 4, &to, 4, sum);
+				if (l4_csum_replace(ctx, off + offsetof(struct icmphdr, checksum),
```
Comment (jschwinger233) [inline bpf/lib/nat.h]: Same as above.

File: bpf/lib/nat.h [hunk 4]
```diff
@@ -1642,17 +1639,17 @@ static __always_inline int snat_v6_rewrite_egress(struct __ctx_buff *ctx,
...
 			if (ctx_store_bytes(ctx, off +
 					    offsetof(struct icmp6hdr,
 						     icmp6_dataun.u_echo.identifier),
 					    &state->to_sport,
 					    sizeof(state->to_sport), 0) < 0)
 				return DROP_WRITE_ERROR;
-			from = tuple->sport;
-			to = state->to_sport;
-			sum = csum_diff(&from, 4, &to, 4, sum);
+			if (l4_csum_replace(ctx, off + offsetof(struct icmp6hdr, icmp6_cksum),
```
Comment (sahid) [inline bpf/lib/nat.h]: Actually I should not do it for icmp6

File: bpf/lib/nat.h (hunk 2 above)
Comment (sahid) [reply bpf/lib/nat.h]: I should not do this at all for icmp6

**Comment by jschwinger233** [comment]
Do you think your change leaves these code dead, as we no longer update the `sum_l4` and `csum.offset` in the ICMP branch?
https://github.com/cilium/cilium/blob/5c94a0dc975bdc067cd8b8886ca60ed694471e1e/bpf/lib/nat.h#L563-L564

**Review by jschwinger233 — APPROVED** [review]
Let's move forward and don't block your following PRs, we can clean some dead code later if there is any.

**Comment by jschwinger233** [comment]
No worries, I should have pointed out the potential issues in the first place.

File: bpf/lib/nat.h [hunk 5]
```diff
@@ -542,18 +542,16 @@ snat_v4_icmp_rewrite_egress_embedded(struct __ctx_buff *ctx,
...
-			from = tuple->sport;
-			to = state->to_sport;
-			flags = 0; /* ICMPv4 has no pseudo-header */
-			sum_l4 = csum_diff(&from, 4, &to, 4, 0);
-			csum.offset = offsetof(struct icmphdr, checksum);
+			if (l4_csum_replace(ctx, inner_l4_off + offsetof(struct icmphdr, checksum),
+					    tuple->sport,
+					    state->to_sport,
+					    sizeof(tuple->sport)) < 0)
+				return DROP_WRITE_ERROR;
```
Comment (jschwinger233) [inline bpf/lib/nat.h]: I'm thinking about returning `DROP_CSUM_L4` rather than `DROP_WRITE_ERROR`, as we did before at https://github.com/cilium/cilium/blob/b93af2b484cc16f05c281a2be227066783322b15/bpf/lib/nat.h#L569

Comment (jschwinger233) [reply bpf/lib/nat.h]: Maybe we can change this returned error in your following PR.

**Review by jschwinger233 — APPROVED** [review]

**Comment by jschwinger233** [comment]
CI failed due to BPF complexity issue lol

File: bpf/lib/nat.h (hunk 5 above)
Comment (sahid) [reply bpf/lib/nat.h]: Yes right It's probably what I should have used.

**Comment by sahid** [comment]
> CI failed due to BPF complexity issue lol

Are you sure, I see all the k8s-*kernel tests green, no?

In all cases if I have to respine for the complexity issue I will use that to fix all the point you have mentioned previously.

**Comment by jschwinger233** [comment]
It's L4LB XDP test: https://github.com/cilium/cilium/actions/runs/5172796260/jobs/9317480797#step:11:2196

File: USERS.md [hunk 6]
```diff
@@ -509,6 +509,12 @@ Users (Alphabetically)
       U: CNI (instead of amazon-vpc-cni-k8s), DefaultDeny(Zero Trust), Hubble, CiliumNetworkPolicy, CiliumClusterwideNetworkPolicy, EKS
       L: https://schubergphilis.com/en
       Q: @stimmerman @shoekstra @mbaumann
+      
+    * N: SIGHUP
```
Comment (ti-mo) [inline USERS.md]: Why this change? Seems to be copied from the block below. :thinking:

**Review by ti-mo — CHANGES_REQUESTED** [review]

File: USERS.md (hunk 6 above)
Comment (sahid) [reply USERS.md]: Hum I made a mistake somewhere during my rebasing... I will fix that

File: bpf/lib/nat.h (hunk 5 above)
Comment (joestringer) [reply bpf/lib/nat.h]: @jschwinger233 @sahid what's the tradeoff, any particular reason to defer that to a subsequent PR? I'm not sure I understand why we don't just address the feedback in the PR that's adding the code. :thinking:

Comment (jschwinger233) [reply bpf/lib/nat.h]: @joestringer That was because Sahid said he had another PR based on this, so I thought we can defer the minor changes to that PR, so as to make things happen fast. But that was before this PR hit the CI failures, since this PR has been here for weeks, I think we should address the issues at this moment. Thanks for pointing out.

Comment (sahid) [reply bpf/lib/nat.h]: @joestringer It's my mistake, I should have fixed both of the suggestions. I wanted to rebase the code and see if by any chance the complexity issue was gone... Sorry @jschwinger233 to have put you in that situation.

All is fixed now.

**Review by ti-mo — APPROVED** [review]

**Comment by sahid** [comment]
The previous PR has been closed du to a lack of activity I created a new one #26453

**Comment by borkmann** [comment]
Heya :wave: Why is this change better than before? Imo, it makes things worse. Now we need an additional helper call whereas before it was not needed and the csum_diff() was actually all inlined.

**Comment by sahid** [comment]
> Heya wave Why is this change better than before? Imo, it makes things worse. Now we need an additional helper call whereas before it was not needed and the csum_diff() was actually all inlined.

:wave:, previously the code was forcing some fields in https://github.com/cilium/cilium/blob/main/bpf/lib/csum.h#L60 and later the block to retrieve `sum` from `sum_l4` a bit obscure (I considered that a bit hacky). I thought I was walking in the good direction changing this part :disappointed_relieved:
